src/core/utils/flash.cpp

```diff
@@ -0,0 +1,266 @@
+        otPlatFlashRead(&GetInstance(), mSwapIndex, 0, &swapMarker, sizeof(swapMarker));
+
+        if (swapMarker == sSwapActive)
+        {
+            break;
+        }
+    }
+
+    for (mSwapUsed = kSwapMarkerSize; mSwapUsed < mSwapSize; mSwapUsed += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, mSwapUsed, &record, sizeof(record));
+        VerifyOrExit(record.IsAddBeginSet());
+    }
+
+exit:
+    return;
+}
+
+otError Flash::Get(uint16_t aKey, int aIndex, uint8_t *aValue, uint16_t *aValueLength) const
+{
+    otError      error       = OT_ERROR_NOT_FOUND;
+    uint16_t     valueLength = 0;
+    int          index       = 0;
+    uint32_t     offset;
+    RecordHeader record;
+
+    for (offset = kSwapMarkerSize; offset < mSwapUsed; offset += record.GetSize())
+    {
+        otPlatFlashRead(&GetInstance(), mSwapIndex, offset, &record, sizeof(record));
+
+        if ((record.GetKey() != aKey) || !record.IsValid())
+        {
+            continue;
+        }
+
+        if (record.IsFirst())
+        {
+            index = 0;
+        }
+
+        if (index == aIndex)
+        {
+            if (aValue && aValueLength)
+            {
+                uint16_t readLength = *aValueLength;
+
+                if (readLength > record.GetLength())
+                {
+                    readLength = record.GetLength();
+                }
+
+                otPlatFlashRead(&GetInstance(), mSwapIndex, offset + sizeof(record), aValue, readLength);
+            }
+
+            valueLength = record.GetLength();
+            error       = OT_ERROR_NONE;
+        }
+
+        index++;
+    }
+
+    if (aValueLength)
+    {
+        *aValueLength = valueLength;
+    }
+
+    return error;
+}
+
+otError Flash::Set(uint16_t aKey, const uint8_t *aValue, uint16_t aValueLength)
+{
+    return Add(aKey, true, aValue, aValueLength);
+}
+
+otError Flash::Add(uint16_t aKey, const uint8_t *aValue, uint16_t aValueLength)
+{
+    bool first = (Get(aKey, 0, NULL, NULL) == OT_ERROR_NOT_FOUND);
+
+    return Add(aKey, first, aValue, aValueLength);
+}
+
+otError Flash::Add(uint16_t aKey, bool aFirst, const uint8_t *aValue, uint16_t aValueLength)
+{
+    otError error = OT_ERROR_NONE;
+    Record  record;
+
+    record.Init(aKey, aFirst);
+    record.SetData(aValue, aValueLength);
+
+    if ((mSwapSize - record.GetSize()) < mSwapUsed)
+    {
+        Swap();
+        VerifyOrExit((mSwapSize - record.GetSize()) >= mSwapUsed, error = OT_ERROR_NO_BUFS);
+    }
+
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, record.GetSize());
+
+    record.SetAddCompleteFlag();
+    otPlatFlashWrite(&GetInstance(), mSwapIndex, mSwapUsed, &record, sizeof(RecordHeader));
+
+    mSwapUsed += record.GetSize();
+
+exit:
+    return error;
+}
+
+bool Flash::DoesValidRecordExist(uint32_t aOffset, uint16_t aKey) const
+{
+    RecordHeader record;
+    bool         rval = false;
+
+    for (; aOffset < mSwapUsed; aOffset += record.GetSize())
```

Comment:
This loop does not guarantee that `otPlatFlashRead` will not read past the end of the swap block. Maybe `aOffset < mSwapSize - sizeof(record)`?